### PR TITLE
An alternative to the exhaustive join enumeration xform

### PR DESCRIPTION
**Pull request by zellerh**
tl;dr

A new xform, CXformExpandNAryJoinDPv2 (DPv2), is introduced, that does the same
work as the existing CXformExpandNAryJoinDP (DP). It is faster for large joins and
works bottom-up instead of top-down.

More details:

The new xform is bottom-up instead of top-down. This is more efficient
for large joins, as it avoids large numbers of hash table lookups.
For small joins we don't see much performance improvements, though.

At most one of these two transforms (DP or DPv2) should be enabled at
a time. By default, the existing xform (DP) is enabled and DPv2 is
disabled. We added a new value "auto1" for the "optimizer_join_order"
guc. Using this new option, the DP xform is disabled and DPv2 is
enabled. DPv2 is experimental at this stage, we do not yet recommend
it for general use.

Example:

```
set optimizer_join_order=auto1;
select ...
```

Which transforms are enabled:

```
optimizer_join_order   DP         DPv2       greedy, query, etc.
--------------------   ---------  ---------- --------------------
exhaustive (default)   enabled    disabled   enabled
auto1                  disabled   enabled    enabled
greedy, query          disabled   disabled   one xform enabled
```

Both the DP and DPv2 transforms usually generate equivalent expressions
after the exploration phase, but the order of these expressions in the MEMO
data structure will generally be different. This can lead to plan differences,
for several reasons:

- Estimated cardinalities may change, since we may be using different join
  orders to compute the estimates, e.g. "(t1 join t2) join t3" vs.
  "(t2 join t3) join t1".
- Transforms that are marked as "ApplyOnce" may see a different binding
  when they are executed.
- When we encounter ties in cost comparison during join enumeration, the two
  transforms may break those ties in different ways.

We have seen clear evidence of the first issue, but no evidence so far for the
second and third issue.

Note that both DP and DPv2 use the same simplified, cardinality-based cost model
for enumerating join orders. This may lead to eliminating join orders that lead
to good partition selectors, to a reduced number of motions, to good use of
indexes, etc. Future work should try to improve this situation, either by
adding more information to the cost model and/or by letting the DPv2
transform generate more alternatives that can then be evaluated by the
actual cost model.

One goal that we have not yet achieved with this commit is to create a transform
that can exhaustively enumerate larger joins (of more than 10 tables) and that
transitions gradually to a greedy approach as the joins get larger. We hope to
do that in a future commit. Some ideas we thought about:

- Reduce the number of cross products that we consider, even for smaller joins.
  (E.g. pass "false" to CJoinOrderDPv2::SearchJoinOrders() to suppress cross joins
  and only pass "true" if we can't generate any joins on a given level).
- Gradually reduce the number of bushy trees for larger joins.
- Gradually reduce the number of linear trees for even larger joins
  (eliminate alternative subsets with higher costs for a given level)
- Do some performance optimizations in the CBitSet class

Data structures:

The DPv2 xform uses an array of levels to enumerate potential join orders.
Index n in this array corresponds to a list of possible n-way joins.
Index 0 is unused, as we are not interested in 0-way joins. Index 1
contains the children of the NAry join. Index 2 contains all enumerated
2-way joins, etc., up to index n.

Each level is again an array of bitsets, with each bitset representing
the set of joined tables. Level 1 has n different bitsets, each with a
single bit set. Level 2 has bitsets with two bits in them, etc. Level n,
the top level, has only one bitset with all n bits set. Together with
each bitset, we also store the CExpression for the cheapest join and
the (cardinality-based) cost of that cheapest join. The top level is
an exception, we store the 10 cheapest joins (in a separate list).

Some performance comparisons between the DP and DPv2 transformations:

Test 1: Simple linear and fully connected join queries

- We tested 3 to 12-way inner joins of very simple, empty tables.
  For linear queries, the join graph is a line, with each table having a
  join predicate with only one or two other tables. Fully connected queries
  have actual or inferred join predicates between any pair of tables.

  Planning times for linear and fully connected joins in msec

  ```
            linear joins                fully-connected
  ----------  ------------------------   -------------------------
  n-way join    DP     DPv2   pct diff    DP     DPv2     pct diff
  ----------  ------  ------  --------   ------  -------  --------
       3          68      97        43       97       84       -13
       4          78      85         9      231      168       -27
       5         145     108       -26      713      264       -63
       6         173     108       -38      945      380       -60
       7         251     161       -36    1,241      558       -55
       8         332     259       -22    1,715      863       -50
       9         628     609        -3    2,501    1,344       -46
      10       1,550   1,359       -12    6,494    2,529       -61
      11       4,838   3,767       -22   27,757    5,265       -81
      12      20,565  12,689       -38   95,899   11,780       -88
 
 ```

  Note: We used set "optimizer_join_order_threshold = 12" to allow the DP and DPv2
  transforms to handle up to 12-way joins (default is 10-way joins).

Test 2: Planning time of TPC-DS queries:

- Total planning time is basically unchanged, within 1%. For individual queries,
  we see some changes in planning time, Q71 planning time improves by 9%,
  Q28 planning time degrades by 19%. Note that these time measurements have
  a certain amount of statistical noise in them.
- Query plans change slightly.
- If we reduce the number of cross products in DPv2, we saw a significant change
  for Q72 in plan performance, due to a bad partition selector. This led us to
  keep producing similar cross joins as for the DP transform, even though this
  hurts planning time very substantially.

Test 3: Workload of about 2,000 real-world queries

- Results are similar to TPC-DS, with no significant changes in planning time
  and some changes in plans, which is expected, based on the previous bullet points.

Co-authored-by: Bhuvnesh Chaudhary <bchaudhary@pivotal.io>
Co-authored-by: Chris Hajas <chajas@pivotal.io>

### CODE REVIEW

File: libgpopt/src/xforms/CXformExpandNAryJoinDPv2.cpp [hunk 1]
```diff
@@ -0,0 +1,145 @@
+//---------------------------------------------------------------------------
+//	Greenplum Database
+//	Copyright (C) 2013 EMC Corp.
```
Comment (hsyuan) [inline libgpopt/src/xforms/CXformExpandNAryJoinDPv2.cpp]: Copyright.

**Comment by zellerh** [comment]
Hi @hsyuan, the original paper is by Pat Selinger et. al.: Access path selection in a relational database management system, https://dl.acm.org/citation.cfm?doid=582095.582099. A good summary is Surajit Chaudhuri: An overview of query optimization in relational systems, https://web.stanford.edu/class/cs345d-01/rl/chaudhuri98.pdf (section on System-R optimizer). Unlike these papers, our dynamic programming transform uses a simplified cost model and it does not consider "interesting orders". I found a description of this simplified approach in a book by Hector Garcia-Molina and Jennifer Widom: Database System Implementation, https://www.amazon.com/gp/product/8131704130.

**Comment by hsyuan** [comment]
Thanks a lot, @zellerh!

File: libgpopt/include/gpopt/xforms/CXformExpandNAryJoinDPv2.h [hunk 2]
```diff
@@ -0,0 +1,90 @@
+//---------------------------------------------------------------------------
+//	Greenplum Database
+//	Copyright (C) 2013 EMC Corp.
```
Comment (bhuvnesh2703) [inline libgpopt/include/gpopt/xforms/CXformExpandNAryJoinDPv2.h]: copyright of pivotal?

File: libgpopt/include/gpopt/xforms/CJoinOrderDPv2.h [hunk 3]
```diff
@@ -0,0 +1,211 @@
+//---------------------------------------------------------------------------
+//	Greenplum Database
+//	Copyright (C) 2013 EMC Corp.
```
Comment (bhuvnesh2703) [inline libgpopt/include/gpopt/xforms/CJoinOrderDPv2.h]: copyright of pivotal?

File: libgpopt/include/gpopt/xforms/CJoinOrderDP.h [hunk 4]
```diff
@@ -98,6 +98,7 @@ namespace gpopt
 			typedef CHashMap<CBitSet, CExpression, UlHashBitSet, FEqualBitSet,
 				CleanupRelease<CBitSet>, CleanupRelease<CExpression> > BitSetToExpressionMap;
 
+
```
Comment (bhuvnesh2703) [inline libgpopt/include/gpopt/xforms/CJoinOrderDP.h]: remove this file from the commit..

File: libgpopt/src/operators/CLogicalNAryJoin.cpp [hunk 5]
```diff
@@ -74,11 +74,12 @@ CLogicalNAryJoin::PxfsCandidates
...
-	(void) xform_set->ExchangeSet(CXform::ExfExpandNAryJoin);
-	(void) xform_set->ExchangeSet(CXform::ExfExpandNAryJoinMinCard);
-	(void) xform_set->ExchangeSet(CXform::ExfExpandNAryJoinDP);
-	(void) xform_set->ExchangeSet(CXform::ExfExpandNAryJoinGreedy);
+    (void) xform_set->ExchangeSet(CXform::ExfSubqNAryJoin2Apply);
+    (void) xform_set->ExchangeSet(CXform::ExfExpandNAryJoin);
+    (void) xform_set->ExchangeSet(CXform::ExfExpandNAryJoinMinCard);
+    (void) xform_set->ExchangeSet(CXform::ExfExpandNAryJoinDP);
+    (void) xform_set->ExchangeSet(CXform::ExfExpandNAryJoinGreedy);
+	(void) xform_set->ExchangeSet(CXform::ExfExpandNAryJoinDPv2);
```
Comment (bhuvnesh2703) [inline libgpopt/src/operators/CLogicalNAryJoin.cpp]: indentation

File: libgpopt/src/xforms/CJoinOrder.cpp [hunk 6]
```diff
@@ -410,6 +410,7 @@ CJoinOrder::PcompCombine
 	SComponent *comp2
 	)
 {
+
```
Comment (bhuvnesh2703) [inline libgpopt/src/xforms/CJoinOrder.cpp]: please remove this file from commit

File: libgpopt/src/xforms/CJoinOrderDP.cpp [hunk 7]
```diff
@@ -572,6 +571,8 @@ CJoinOrderDP::PexprBestJoinOrderDP
 		pexprResult = m_pexprDummy;
 	}
 
+
```
Comment (bhuvnesh2703) [inline libgpopt/src/xforms/CJoinOrderDP.cpp]: please remove this file from the commit.

File: libgpopt/src/xforms/CJoinOrderDPv2.cpp [hunk 8]
```diff
@@ -0,0 +1,786 @@
+//---------------------------------------------------------------------------
+//	Greenplum Database
+//	Copyright (C) 2013 EMC Corp.
```
Comment (bhuvnesh2703) [inline libgpopt/src/xforms/CJoinOrderDPv2.cpp]: pivotal copyright?

File: libgpopt/src/xforms/CJoinOrderDPv2.cpp [hunk 9]
```diff
@@ -0,0 +1,786 @@
...
+	SComponentInfo *left_child,
+	SComponentInfo *right_child,
+	BOOL allow_cross_joins
+	)
+{
+	CExpression *scalar_expr = PexprPred(left_child->component, right_child->component);
+
+	if (NULL == scalar_expr)
+	{
+		if (!allow_cross_joins)
```
Comment (bhuvnesh2703) [inline libgpopt/src/xforms/CJoinOrderDPv2.cpp]: do we need this now? since we always consider cross joins?

File: libgpopt/src/xforms/CJoinOrderDPv2.cpp [hunk 10]
```diff
@@ -0,0 +1,786 @@
...
+		// entry to avoid duplicate join combinations
+		// i.e a join b and b join a, just try one
+		// commutativity will take care of the other
+		ULONG other_pair_start_id = 0;
+		if (join_pair_components == other_join_pair_components)
+			other_pair_start_id = join_pair_id + 1;
+
+		for (ULONG other_pair_id = other_pair_start_id; other_pair_id < other_join_pairs_size; other_pair_id++)
+		{
+			CBitSet *best_join_bitset = GPOS_NEW(m_mp) CBitSet(m_mp, *left_bitset);
```
Comment (bhuvnesh2703) [inline libgpopt/src/xforms/CJoinOrderDPv2.cpp]: not the best_join_bitset, but join_bitset.. as its not the best one..

File: libgpopt/src/xforms/CJoinOrderDPv2.cpp [hunk 11]
```diff
@@ -0,0 +1,786 @@
...
+			result_map->Insert(join_bitset_entry, join_exprs);
+		}
+		else
+		{
+			AddExprs(candidate_join_exprs, existing_join_exprs);
+		}
+	}
+}
+
+CJoinOrderDPv2::BitSetToExpressionArrayMap *
```
Comment (bhuvnesh2703) [inline libgpopt/src/xforms/CJoinOrderDPv2.cpp]: add comments please

File: libgpopt/src/xforms/CJoinOrderDPv2.cpp [hunk 12]
```diff
@@ -0,0 +1,786 @@
...
+//		CJoinOrderDPv2::PExprExpand
+//
+//	@doc:
+//		Main driver for join order enumeration, called by xform
+//
+//---------------------------------------------------------------------------
+CExpression *
+CJoinOrderDPv2::PexprExpand()
+{
+	ComponentInfoArray *non_join_vertex_component_infos = GPOS_NEW(m_mp) ComponentInfoArray(m_mp);
```
Comment (bhuvnesh2703) [inline libgpopt/src/xforms/CJoinOrderDPv2.cpp]: non_join_vertex is bit cryptic, could we use base_relation_component_infos or something similar?

File: libgpopt/src/xforms/CJoinOrderDPv2.cpp [hunk 13]
```diff
@@ -0,0 +1,786 @@
...
+		ULONG previous_level = current_join_level - 1;
+		ComponentInfoArray *prev_lev_comps = (*m_join_levels)[previous_level];
+		// build linear "current_join_level" joins, with a "previous_level"-way join on one
+		// side and a non-join vertex on the other side
+		BitSetToExpressionArrayMap *bitset_join_exprs_map = SearchJoinOrders(prev_lev_comps, non_join_vertex_component_infos, true);
+		// build bushy trees - joins between two other joins
+		BitSetToExpressionArrayMap *bitset_bushy_join_exprs_map = SearchBushyJoinOrders(current_join_level);
+
+		// Commented out for now. Enable this if/when we want to pass "false"
+		// to the "allow_cross_joins" parameter in SearchJoinOrders
```
Comment (bhuvnesh2703) [inline libgpopt/src/xforms/CJoinOrderDPv2.cpp]: do we plan to remove it before merging?

File: libgpopt/src/xforms/CXform.cpp [hunk 14]
```diff
@@ -272,12 +273,37 @@ CBitSet *CXform::PbsJoinOrderOnGreedyXforms
...
 	(void) pbs->ExchangeSet(GPOPT_DISABLE_XFORM_TF(CXform::ExfExpandNAryJoinDP));
+	(void) pbs->ExchangeSet(GPOPT_DISABLE_XFORM_TF(CXform::ExfExpandNAryJoinDPv2));
 	(void) pbs->ExchangeSet(GPOPT_DISABLE_XFORM_TF(CXform::ExfJoinAssociativity));
 	(void) pbs->ExchangeSet(GPOPT_DISABLE_XFORM_TF(CXform::ExfJoinCommutativity));
 
 	return pbs;
 }
 
+CBitSet *CXform::PbsJoinOrderOnExhaustiveXforms
+(
```
Comment (bhuvnesh2703) [inline libgpopt/src/xforms/CXform.cpp]: indentation

File: libgpopt/src/xforms/CXform.cpp [hunk 15]
```diff
@@ -272,12 +273,37 @@ CBitSet *CXform::PbsJoinOrderOnGreedyXforms
...
+{
+	CBitSet *pbs = GPOS_NEW(mp) CBitSet(mp, EopttraceSentinel);
+
+	(void) pbs->ExchangeSet(GPOPT_DISABLE_XFORM_TF(CXform::ExfExpandNAryJoinDPv2));
+
+	return pbs;
+}
+
+CBitSet *CXform::PbsJoinOrderOnAuto1Xforms
+(
```
Comment (bhuvnesh2703) [inline libgpopt/src/xforms/CXform.cpp]: indentation

File: libgpopt/src/xforms/CXformExpandNAryJoinDPv2.cpp [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+	CJoinOrderDPv2 jodp(mp, pdrgpexpr, pdrgpexprPreds);
+	jodp.PexprExpand();
+
+	// Retrieve top K join orders from jodp and add as alternatives
+	const ULONG UlTopKJoinOrders = jodp.PdrgpexprTopK()->Size();
+	for (ULONG ul = 0; ul < UlTopKJoinOrders; ul++)
+	{
+		CExpression *pexprJoinOrder = (*jodp.PdrgpexprTopK())[ul];
+		pexprJoinOrder->AddRef();
+		pxfres->Add(pexprJoinOrder);
```
Comment (bhuvnesh2703) [inline libgpopt/src/xforms/CXformExpandNAryJoinDPv2.cpp]: we should normalize the join expression before inserting?

File: scripts/fix_mdps.py [hunk 17]
```diff
@@ -43,7 +43,7 @@ def replacePlanSize(filename, actual, expected):
     with open(filename, 'w') as fp:
         fp.write(filedata)
 
-def replaceTextBetween(filename, innerContent):
+def replacePlanInFile(filename, innerContent):
```
Comment (bhuvnesh2703) [inline scripts/fix_mdps.py]: is this intentional?

File: libgpopt/src/xforms/CJoinOrderDPv2.cpp [hunk 18]
```diff
@@ -0,0 +1,786 @@
...
+		existing_join_exprs->Append(join_expr);
+	}
+	else
+	{
+		CExpressionArray *exprs = GPOS_NEW(m_mp) CExpressionArray(m_mp);
+		exprs->Append(join_expr);
+		join_bitset->AddRef();
+		BOOL success = map->Insert(join_bitset, exprs);
+		if (!success)
+			raise(1);
```
Comment (bhuvnesh2703) [inline libgpopt/src/xforms/CJoinOrderDPv2.cpp]: please use GPOS_RAISE

File: libgpopt/src/xforms/CJoinOrderDPv2.cpp (hunk 9 above)
Comment (zellerh) [reply libgpopt/src/xforms/CJoinOrderDPv2.cpp]: Ok, I'll remove the allow_cross_join parameters from SearchJoinOrders() and GetJoinExpr().

File: libgpopt/src/xforms/CJoinOrderDPv2.cpp (hunk 18 above)
Comment (zellerh) [reply libgpopt/src/xforms/CJoinOrderDPv2.cpp]: Done.

File: libgpopt/src/xforms/CJoinOrderDPv2.cpp (hunk 10 above)
Comment (zellerh) [reply libgpopt/src/xforms/CJoinOrderDPv2.cpp]: Done.

File: libgpopt/src/xforms/CJoinOrderDPv2.cpp (hunk 11 above)
Comment (zellerh) [reply libgpopt/src/xforms/CJoinOrderDPv2.cpp]: Done.

File: libgpopt/src/xforms/CJoinOrderDPv2.cpp (hunk 12 above)
Comment (zellerh) [reply libgpopt/src/xforms/CJoinOrderDPv2.cpp]: We spent some time thinking about this term and asked several people. @vraghavan78 suggested this term and we liked it. These components could be base relations or they could be aggregates, unions or other expressions, so I'm a bit hesitant to call them base relations.

File: libgpopt/src/xforms/CJoinOrderDPv2.cpp (hunk 13 above)
Comment (zellerh) [reply libgpopt/src/xforms/CJoinOrderDPv2.cpp]: Sure, I'll remove it.

File: libgpopt/src/xforms/CXform.cpp (hunk 14 above)
Comment (zellerh) [reply libgpopt/src/xforms/CXform.cpp]: Done.

File: libgpopt/src/xforms/CXform.cpp (hunk 15 above)
Comment (zellerh) [reply libgpopt/src/xforms/CXform.cpp]: Done.

File: libgpopt/src/xforms/CXformExpandNAryJoinDPv2.cpp (hunk 1 above)
Comment (zellerh) [reply libgpopt/src/xforms/CXformExpandNAryJoinDPv2.cpp]: Done.

File: libgpopt/src/xforms/CXformExpandNAryJoinDPv2.cpp (hunk 16 above)
Comment (zellerh) [reply libgpopt/src/xforms/CXformExpandNAryJoinDPv2.cpp]: Thanks. We have the same issue in the old DP2 transformation, I'll change it there as well.

File: scripts/fix_mdps.py (hunk 17 above)
Comment (zellerh) [reply scripts/fix_mdps.py]: Yes, because the method really can only replace the plan.

File: libgpopt/src/xforms/CJoinOrderDPv2.cpp [hunk 19]
```diff
@@ -0,0 +1,783 @@
...
+		for (ULONG ul = 0; ul < arity - 1; ul++)
+		{
+			CExpression *pexprChild = (*pexpr)[ul];
+
+			// call function recursively to find child cost
+			// TODO: This will count cost of leaves twice
+			dCost = dCost + DCost(pexprChild);
+			DeriveStats(pexprChild);
+			rgdRows[ul] = pexprChild->Pstats()->Rows().Get();
+			// TODO: Consider width of the rows
```
[the remark this answers is not in the document]
Comment (zellerh) [reply libgpopt/src/xforms/CJoinOrderDPv2.cpp]: Removed the TODOs for now and replaced them with notes. Maybe we can consider the width in some future enhancement.

File: libgpopt/include/gpopt/xforms/CJoinOrderDPv2.h [hunk 20]
```diff
@@ -0,0 +1,211 @@
...
+					{
+						this->component = component;
+						this->best_expr = best_expr;
+					}
+
+					~SComponentInfo()
+					{
+						if (NULL != component)
+						{
+							component->Release();
```
[the remark this answers is not in the document]
Comment (zellerh) [reply libgpopt/include/gpopt/xforms/CJoinOrderDPv2.h]: Done.

File: libgpopt/include/gpopt/xforms/CJoinOrderDPv2.h [hunk 21]
```diff
@@ -0,0 +1,211 @@
...
+			//		Struct containing a component, its best expression, and cost
+			//
+			//---------------------------------------------------------------------------
+			struct SComponentInfo : public CRefCount
+				{
+					CBitSet *component;
+					CExpression *best_expr;
+					CDouble cost;
+
+					SComponentInfo() : cost(0.0)
```
[the remark this answers is not in the document]
Comment (zellerh) [reply libgpopt/include/gpopt/xforms/CJoinOrderDPv2.h]: Done.

File: libgpopt/src/xforms/CJoinOrderDPv2.cpp [hunk 22]
```diff
@@ -0,0 +1,783 @@
...
+{
+#ifdef GPOS_DEBUG
+	// in optimized build, we flush-down memory pools without leak checking,
+	// we can save time in optimized build by skipping all de-allocations here,
+	// we still have all de-llocations enabled in debug-build to detect any possible leaks
+	m_topKOrders->Release();
+	m_topKCosts->Release();
+	m_pexprDummy->Release();
+	m_join_levels->Release();
+	CAutoTrace at(m_mp);
```
[the remark this answers is not in the document]
Comment (zellerh) [reply libgpopt/src/xforms/CJoinOrderDPv2.cpp]: Thanks, good point, it's now removed.

File: libgpopt/src/operators/CLogicalNAryJoin.cpp [hunk 23]
```diff
@@ -74,11 +74,12 @@ CLogicalNAryJoin::PxfsCandidates
 {
 	CXformSet *xform_set = GPOS_NEW(mp) CXformSet(mp);
 	
-	(void) xform_set->ExchangeSet(CXform::ExfSubqNAryJoin2Apply);
-	(void) xform_set->ExchangeSet(CXform::ExfExpandNAryJoin);
-	(void) xform_set->ExchangeSet(CXform::ExfExpandNAryJoinMinCard);
-	(void) xform_set->ExchangeSet(CXform::ExfExpandNAryJoinDP);
-	(void) xform_set->ExchangeSet(CXform::ExfExpandNAryJoinGreedy);
+    (void) xform_set->ExchangeSet(CXform::ExfSubqNAryJoin2Apply);
```
Comment (ashuka24) [inline libgpopt/src/operators/CLogicalNAryJoin.cpp]: tab spacing

File: data/dxl/minidump/EagerAggMaxWithNestedLoop.mdp [hunk 24]
```diff
@@ -61,7 +61,7 @@
                 </dxl:CostParams>
             </dxl:CostModelConfig>
             <dxl:Hint ArrayExpansionThreshold="100" BroadcastThreshold="10000000" EnforceConstraintsOnDML="false" JoinArityForAssociativityCommutativity="18" JoinOrderDynamicProgThreshold="10" MinNumOfPartsToRequireSortOnInsert="2147483647"/>
-            <dxl:TraceFlags Value="102074,102120,103001,103014,103015,103022,103027,103030,104003,104004,104005,105000,106000"/>
+            <dxl:TraceFlags Value="102074,102146,102120,103001,103014,103015,103022,103027,102146,103030,104003,104004,104005,105000,106000"/>
```
[the remark this answers is not in the document]
Comment (zellerh) [reply data/dxl/minidump/EagerAggMaxWithNestedLoop.mdp]: Thanks, good catch!

File: libgpopt/src/operators/CLogicalNAryJoin.cpp (hunk 23 above)
Comment (zellerh) [reply libgpopt/src/operators/CLogicalNAryJoin.cpp]: Fixed.

File: libgpopt/src/xforms/CXformExpandNAryJoinDP.cpp [hunk 25]
```diff
@@ -142,8 +142,9 @@ CXformExpandNAryJoinDP::Transform
 			CExpression *pexprJoinOrder = (*jodp.PdrgpexprTopK())[ul];
 			if (pexprJoinOrder != pexprResult)
 			{
-				pexprJoinOrder->AddRef();
-				pxfres->Add(pexprJoinOrder);
+				CExpression *pexprNormalizedTopK = CNormalizer::PexprNormalize(mp, pexprJoinOrder);
```
Comment (ashuka24) [inline libgpopt/src/xforms/CXformExpandNAryJoinDP.cpp]: why do we normalize here?

Comment (zellerh) [reply libgpopt/src/xforms/CXformExpandNAryJoinDP.cpp]: @bhuvnesh2703 commented on the new transformation that we should normalize results before returning them. This is also done for the best result on line 135 above. However, I found many plan differences with this change, so will be backing it out and leaving it for a future PR. This change is not directly related to our new transformation.
